src/vctrs.h

```diff
@@ -169,6 +169,57 @@ enum vctrs_type2 {
   vctrs_type2_scalar_scalar
 };
 
+enum vctrs_s3_type2 {
+  vctrs_s3_type2_null_bare_factor,
+  vctrs_s3_type2_null_bare_ordered,
+  vctrs_s3_type2_null_unknown,
+
+  vctrs_s3_type2_logical_bare_factor,
```

Comment:
So ive made `enum vctrs_s3_type2` include all of the combinations because `enum vctrs_type2` does as well. I think this is mainly useful for testing, because we have that double for loop test that cycles through all of the possible combinations and ensures that we have correctly specified everything. It is currently easy to add new types to the tests because you just add them to the R objects `base_s3_empty_types` or `base_empty_types` and the loop runs through them automatically. I'm a bit worried that if i reduce down to only what is actually used then that will get harder to test consistently.

Right now `type2-dispatch.c` and `type2-dispatch-extension.c` mirror each other nicely because both enumerate all possibilities. I think if you understand one then you can fairly easily understand the other

Reply:
The automatic check is a good point. I still think it's better to keep it simple in this case, especially since the number of combinations is potentially unbounded. But no worries if you prefer to keep the combinations, this will be replaced by another dispatch mechanism anyway.

Reply:
Let's leave for now